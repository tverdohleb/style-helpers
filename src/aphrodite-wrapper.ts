// tslint:disable-next-line:no-submodule-imports
import * as aphrodite from 'aphrodite/no-important';

// tslint:disable-next-line:no-duplicate-imports
import {
  StyleDeclaration,
  StyleDeclarationValue,
  CSSPropertiesComplete,
  CSSPropertiesLossy,
  CSSProperties,
  // tslint:disable-next-line:no-submodule-imports
} from 'aphrodite/no-important';

import { StyleSheetTestUtilsStatic, StyleSheetServerStatic } from './aphrodite.interface';

const hasParentRef = (selector: string): boolean => selector.indexOf('&') !== -1;
const parentRegExp = /&/g;

const nestSelectorHandler = (
  selector: string,
  baseSelector: string,
  generateCSS: (selector: string) => string,
): string | null => {
  if (hasParentRef(selector)) {
    return generateCSS(selector.replace(parentRegExp, baseSelector));
  }
  return null;
};

const globalSelectorHandler = (
  selector: string,
  baseSelector: string,
  generateSubtreeStyles: (selector: string) => string,
): string | null => {
  if (selector[0] !== '*') {
    return null;
  }
  return generateSubtreeStyles(selector.slice(1));
};

const extendedAphrodite = aphrodite.StyleSheet.extend([
  { selectorHandler: globalSelectorHandler },
  { selectorHandler: nestSelectorHandler },
]);
const { StyleSheet } = extendedAphrodite;
const css = extendedAphrodite.css.bind(extendedAphrodite);
const StyleSheetServer: StyleSheetServerStatic = extendedAphrodite.StyleSheetServer;
const StyleSheetTestUtils: StyleSheetTestUtilsStatic = extendedAphrodite.StyleSheetTestUtils;

export {
  StyleSheet,
  css,
  CSSPropertiesLossy,
  CSSProperties,
  StyleDeclaration,
  StyleDeclarationValue,
  CSSPropertiesComplete,
  StyleSheetServer,
  StyleSheetTestUtils,
};
