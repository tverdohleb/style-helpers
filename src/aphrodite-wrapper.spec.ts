import { StyleSheet, css, StyleSheetTestUtils } from './aphrodite-wrapper';

import { JSDOM } from 'jsdom';
import asap from 'asap';

beforeEach(() => {
  // Object.assign(global, { document: new JSDOM('').window.document });
  (global as any).document = new JSDOM('').window.document;
  StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
});

afterEach(() => {
  (global as any).document.close();
  (global as any).document = undefined;
});

const getSheetText = (sheet: CSSStyleSheet): string => {
  let allRules = '';
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < sheet.cssRules.length; i++) {
    allRules += sheet.cssRules[i].cssText + ' ';
  }
  return allRules;
};

const getStyleText = (): string => {
  const styleTags = (global as any).document.getElementsByTagName('style');
  const lastTag = styleTags[styleTags.length - 1];
  return getSheetText(lastTag.sheet);
};

describe('Aphrodite wrapper', () => {
  it('Nested styles', done => {
    const styleSheet = StyleSheet.create({
      parentStyle: {
        '& .childStyle': {
          position: 'relative',
        },
        display: 'block',
      },
    });

    css(styleSheet.parentStyle);

    asap(() => {
      const styleText = getStyleText();
      expect(styleText).toMatchSnapshot();
      done();
    });
  });

  it('Global styles', done => {
    const styleSheet = StyleSheet.create({
      globals: {
        '* p': {
          display: 'block',
        },
      },
    });

    css(styleSheet.globals);

    asap(() => {
      const styleText = getStyleText();
      expect(styleText).toMatchSnapshot();
      done();
    });
  });
});
