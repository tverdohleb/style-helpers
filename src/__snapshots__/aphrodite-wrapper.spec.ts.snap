// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aphrodite wrapper Global styles 1`] = `"p {display: block;} "`;

exports[`Aphrodite wrapper Nested styles 1`] = `".parentStyle_10lk2fs {display: block;} .parentStyle_10lk2fs .childStyle {position: relative;} "`;
